packages/gator-permissions-snap/src/core/permissionHandler.ts

```diff
@@ -196,8 +232,57 @@ export class PermissionHandler<
       updateContext: (args: { updatedContext: TContext }) => Promise<void>;
     }) => {
       let currentContext = initialContext;
-      const rerender = async () =>
+      const rerender = async () => {
         await updateContext({ updatedContext: currentContext });
+      };
+
+      // fetchAccountBalanceCallCounter is used to cancel any previously executed instances of this function.
+      let fetchAccountBalanceCallCounter = 0;
+      const fetchAccountBalance = async (context: TContext) => {
+        fetchAccountBalanceCallCounter += 1;
+        const currentFetchAccountBalanceCallCounter =
+          fetchAccountBalanceCallCounter;
+
+        const { address } = fromCaip10Address(context.accountAddressCaip10);
+
+        const { assetAddress, chainId } = fromCaip19Address(
+          context.tokenAddressCaip19,
+        );
+
+        const { balance, decimals } =
+          await this.#tokenMetadataService.getTokenBalanceAndMetadata({
+            chainId,
+            account: address,
+            assetAddress,
+          });
+
+        if (
+          currentFetchAccountBalanceCallCounter ===
+          fetchAccountBalanceCallCounter
+        ) {
+          this.#tokenBalance = formatUnits({ value: balance, decimals });
+
+          rerender().catch((error) => logger.error(error));
+
+          const fiatBalance =
+            await this.#tokenPricesService.getCryptoToFiatConversion(
+              initialContext.tokenAddressCaip19,
+              bigIntToHex(balance),
+              initialContext.tokenMetadata.decimals,
+            );
+
+          if (
+            currentFetchAccountBalanceCallCounter ===
+            fetchAccountBalanceCallCounter
+          ) {
+            this.#tokenBalanceFiat = fiatBalance;
+            rerender().catch((error) => logger.error(error));
+          }
+        }
+      };
+
+      // we explicitly don't await this as it's a background process that will re-render the UI (twice) once it is complete
+      fetchAccountBalance(currentContext).catch(console.error);
```

Comment:
I think maybe we should use `logger.error` instead of `console.error` here for consistency and better log management.

 ```ts
 fetchAccountBalance(currentContext).catch((error) => {
     logger.error('Failed to fetch account balance', error);
     // Consider showing the error status to the user.
 });
 ```

Reply:
Oops! I started doing that, and didn't catch all of the instances. I've changed how this works a little bit to correctly await the rerenders. That leaves a single floating promise, when we catch and log correctly.

Regarding presenting this to the user, there's a ticket open for handling token price errors. Presently we catch within the token price service, and return `$<-->` as a placeholder. I think we should throw and handle the errors here, where we can surface a more meaningful message.